### PR TITLE
Fix the remaining Prefast warnings in CPU EP

### PR DESCRIPTION
**Pull request by snnn**
### Description

Fix the remaining Prefast warnings in CPU EP. When this PR is merged, I will be able to enforce the check in our CI build pipelines.

1. PREfast Note C26409 - File: onnxruntime/core/session/abi_session_options.cc. Line: 25. Column 31. 
Tool: PREfast: Rule: C26409 
Avoid calling new and delete explicitly, use std::make_unique<T> instead (r.11).
2. PREfast Note C26409 - File: onnxruntime/core/session/abi_session_options.cc. Line: 36. Column 31. 
Tool: PREfast: Rule: C26409 
Avoid calling new and delete explicitly, use std::make_unique<T> instead (r.11).
3. PREfast Note C26409 - File: onnxruntime/core/session/onnxruntime_c_api.cc. Line: 833. Column 31. 
Tool: PREfast: Rule: C26409 
Avoid calling new and delete explicitly, use std::make_unique<T> instead (r.11).
4. PREfast Note C26409 - File: onnxruntime/core/session/onnxruntime_c_api.cc. Line: 873. Column 26. 
Tool: PREfast: Rule: C26409 
Avoid calling new and delete explicitly, use std::make_unique<T> instead (r.11).
5. PREfast Note C26409 - File: onnxruntime/core/session/onnxruntime_c_api.cc. Line: 988. Column 28. 
Tool: PREfast: Rule: C26409 
Avoid calling new and delete explicitly, use std::make_unique<T> instead (r.11).
6. PREfast Note C26409 - File: onnxruntime/core/session/onnxruntime_c_api.cc. Line: 1328. Column 63. 
Tool: PREfast: Rule: C26409 
Avoid calling new and delete explicitly, use std::make_unique<T> instead (r.11).
7. PREfast Note C26409 - File: onnxruntime/core/session/onnxruntime_c_api.cc. Line: 1988. Column 37. 
Tool: PREfast: Rule: C26409 
Avoid calling new and delete explicitly, use std::make_unique<T> instead (r.11).
8. PREfast Note C26409 - File: onnxruntime/core/session/onnxruntime_c_api.cc. Line: 2019. Column 22. 
Tool: PREfast: Rule: C26409 
Avoid calling new and delete explicitly, use std::make_unique<T> instead (r.11).
9. PREfast Note C26409 - File: onnxruntime/core/session/onnxruntime_c_api.cc. Line: 2022. Column 16. 
Tool: PREfast: Rule: C26409 
Avoid calling new and delete explicitly, use std::make_unique<T> instead (r.11).
10. PREfast Note C26409 - File: onnxruntime/core/session/onnxruntime_c_api.cc. Line: 2046. Column 7. 
Tool: PREfast: Rule: C26409 
Avoid calling new and delete explicitly, use std::make_unique<T> instead (r.11).
11. PREfast Note C26409 - File: onnxruntime/core/session/onnxruntime_c_api.cc. Line: 2048. Column 5. 
Tool: PREfast: Rule: C26409 
Avoid calling new and delete explicitly, use std::make_unique<T> instead (r.11).
12. PREfast Note C26409 - File: onnxruntime/core/session/onnxruntime_c_api.cc. Line: 2142. Column 25. 
Tool: PREfast: Rule: C26409 
Avoid calling new and delete explicitly, use std::make_unique<T> instead (r.11).
13. PREfast Note C26409 - File: onnxruntime/core/session/onnxruntime_c_api.cc. Line: 2188. Column 85. 
Tool: PREfast: Rule: C26409 
Avoid calling new and delete explicitly, use std::make_unique<T> instead (r.11).
14. PREfast Note C26409 - File: onnxruntime/core/session/standalone_op_invoker.cc. Line: 535. Column 5. 
Tool: PREfast: Rule: C26409 
Avoid calling new and delete explicitly, use std::make_unique<T> instead (r.11).
15. PREfast Note C26814 - File: onnxruntime/test/contrib_ops/attention_op_test.cc. Line: 362. Column 29. 
Tool: PREfast: Rule: C26814 
The const variable 'disable_rocm' can be computed at compile-time. Consider using constexpr (con.5).
16. PREfast Note C26814 - File: onnxruntime/test/contrib_ops/attention_op_test.cc. Line: 445. Column 28. 
Tool: PREfast: Rule: C26814 
The const variable 'disable_cpu' can be computed at compile-time. Consider using constexpr (con.5).
17. PREfast Note C26814 - File: onnxruntime/test/contrib_ops/attention_op_test.cc. Line: 447. Column 29. 
Tool: PREfast: Rule: C26814 
The const variable 'disable_rocm' can be computed at compile-time. Consider using constexpr (con.5).
18. PREfast Note C26814 - File: onnxruntime/test/contrib_ops/attention_op_test.cc. Line: 2037. Column 28. 
Tool: PREfast: Rule: C26814 
The const variable 'disable_cpu' can be computed at compile-time. Consider using constexpr (con.5).
19. PREfast Note C26814 - File: onnxruntime/test/contrib_ops/attention_op_test.cc. Line: 2038. Column 29. 
Tool: PREfast: Rule: C26814 
The const variable 'disable_cuda' can be computed at compile-time. Consider using constexpr (con.5).
20. PREfast Note C26814 - File: onnxruntime/test/contrib_ops/attention_op_test.cc. Line: 2039. Column 29. 
Tool: PREfast: Rule: C26814 
The const variable 'disable_rocm' can be computed at compile-time. Consider using constexpr (con.5).
21. PREfast Note C26814 - File: onnxruntime/test/framework/tunable_op_test.cc. Line: 110. Column 17. 
Tool: PREfast: Rule: C26814 
The const variable 'a' can be computed at compile-time. Consider using constexpr (con.5).
22. PREfast Note C26814 - File: onnxruntime/test/framework/tunable_op_test.cc. Line: 111. Column 17. 
Tool: PREfast: Rule: C26814 
The const variable 'b' can be computed at compile-time. Consider using constexpr (con.5).
23. PREfast Note C26400 - File: onnxruntime/test/framework/tunable_op_test.cc. Line: 373. Column 45. 
Tool: PREfast: Rule: C26400 
Do not assign the result of an allocation or a function call with an owner<T> return value to a raw pointer, use owner<T> instead (i.11).
24. PREfast Note C26814 - File: onnxruntime/test/onnx/main.cc. Line: 69. Column 29. 
Tool: PREfast: Rule: C26814 
The const variable 'absolute' can be computed at compile-time. Consider using constexpr (con.5).
25. PREfast Note C6262 - File: onnxruntime/test/optimizer/avx2_weight_s8_to_u8_test.cc. Line: 84. Column 1. 
Tool: PREfast: Rule: C6262 
Function uses '16640' bytes of stack.  Consider moving some data to heap.
26. PREfast Note C6262 - File: onnxruntime/test/optimizer/avx2_weight_s8_to_u8_test.cc. Line: 196. Column 1. 
Tool: PREfast: Rule: C6262 
Function uses '16592' bytes of stack.  Consider moving some data to heap.
27. PREfast Note C6262 - File: onnxruntime/test/optimizer/avx2_weight_s8_to_u8_test.cc. Line: 450. Column 1. 
Tool: PREfast: Rule: C6262 
Function uses '16548' bytes of stack.  Consider moving some data to heap.
28. PREfast Note C6262 - File: onnxruntime/test/optimizer/avx2_weight_s8_to_u8_test.cc. Line: 563. Column 1. 
Tool: PREfast: Rule: C6262 
Function uses '16524' bytes of stack.  Consider moving some data to heap.
29. PREfast Note C6262 - File: onnxruntime/test/optimizer/avx2_weight_s8_to_u8_test.cc. Line: 698. Column 1. 
Tool: PREfast: Rule: C6262 
Function uses '16824' bytes of stack.  Consider moving some data to heap.
30. PREfast Note C26814 - File: onnxruntime/test/optimizer/graph_transform_test.cc. Line: 3874. Column 25. 
Tool: PREfast: Rule: C26814 
The const variable 'alpha' can be computed at compile-time. Consider using constexpr (con.5).
31. PREfast Note C26814 - File: onnxruntime/test/optimizer/graph_transform_test.cc. Line: 3907. Column 25. 
Tool: PREfast: Rule: C26814 
The const variable 'alpha' can be computed at compile-time. Consider using constexpr (con.5).
32. PREfast Note C6011 - File: onnxruntime/test/optimizer/graph_transform_test.cc. Line: 5796. Column 13. 
Tool: PREfast: Rule: C6011 
Dereferencing NULL pointer 'add_initializer'. 
33. PREfast Note C6011 - File: onnxruntime/test/optimizer/graph_transform_test.cc. Line: 5799. Column 20. 
Tool: PREfast: Rule: C6011 
Dereferencing NULL pointer 'clip_min_initializer'. 
34. PREfast Note C6011 - File: onnxruntime/test/optimizer/graph_transform_test.cc. Line: 5802. Column 20. 
Tool: PREfast: Rule: C6011 
Dereferencing NULL pointer 'clip_max_initializer'. 
35. PREfast Note C6011 - File: onnxruntime/test/optimizer/graph_transform_test.cc. Line: 5805. Column 20. 
Tool: PREfast: Rule: C6011 
Dereferencing NULL pointer 'sub_initializer'. 
36. PREfast Note C6011 - File: onnxruntime/test/optimizer/graph_transform_test.cc. Line: 5808. Column 20. 
Tool: PREfast: Rule: C6011 
Dereferencing NULL pointer 'mul_initializer'. 
37. PREfast Note C26814 - File: onnxruntime/test/optimizer/graph_transform_test.cc. Line: 5793. Column 39. 
Tool: PREfast: Rule: C26814 
The const variable 'require_constant' can be computed at compile-time. Consider using constexpr (con.5).
38. PREfast Note C6011 - File: onnxruntime/test/optimizer/graph_transform_test.cc. Line: 6037. Column 18. 
Tool: PREfast: Rule: C6011 
Dereferencing NULL pointer 'add_initializer'. 
39. PREfast Note C6011 - File: onnxruntime/test/optimizer/graph_transform_test.cc. Line: 6139. Column 11. 
Tool: PREfast: Rule: C6011 
Dereferencing NULL pointer 'mul_initializer'. 
40. PREfast Note C6011 - File: onnxruntime/test/optimizer/graph_transform_test.cc. Line: 6145. Column 18. 
Tool: PREfast: Rule: C6011 
Dereferencing NULL pointer 'sub_initializer'. 
41. PREfast Note C26451 - File: onnxruntime/contrib_ops/cpu/bert/attention_base.cc. Line: 378. Column 64. 
Tool: PREfast: Rule: C26451 
Arithmetic overflow: Using operator '+' on a 4 byte value and then casting the result to a 8 byte value. Cast the value to the wider type before calling operator '+' to avoid overflow (io.2).
42. PREfast Note C26451 - File: onnxruntime/contrib_ops/cpu/bert/attention_cpu_base.h. Line: 235. Column 43. 
Tool: PREfast: Rule: C26451 
Arithmetic overflow: Using operator '*' on a 4 byte value and then casting the result to a 8 byte value. Cast the value to the wider type before calling operator '*' to avoid overflow (io.2).
43. PREfast Note C26451 - File: onnxruntime/contrib_ops/cpu/bert/attention_cpu_base.h. Line: 242. Column 29. 
Tool: PREfast: Rule: C26451 
Arithmetic overflow: Using operator '*' on a 4 byte value and then casting the result to a 8 byte value. Cast the value to the wider type before calling operator '*' to avoid overflow (io.2).
44. PREfast Note C26451 - File: onnxruntime/contrib_ops/cpu/bert/attention_cpu_base.h. Line: 210. Column 58. 
Tool: PREfast: Rule: C26451 
Arithmetic overflow: Using operator '*' on a 4 byte value and then casting the result to a 8 byte value. Cast the value to the wider type before calling operator '*' to avoid overflow (io.2).
45. PREfast Note C26451 - File: onnxruntime/contrib_ops/cpu/bert/attention_cpu_base.h. Line: 211. Column 59. 
Tool: PREfast: Rule: C26451 
Arithmetic overflow: Using operator '*' on a 4 byte value and then casting the result to a 8 byte value. Cast the value to the wider type before calling operator '*' to avoid overflow (io.2).
46. PREfast Note C26451 - File: onnxruntime/contrib_ops/cpu/bert/attention_cpu_base.h. Line: 216. Column 15. 
Tool: PREfast: Rule: C26451 
Arithmetic overflow: Using operator '*' on a 4 byte value and then casting the result to a 8 byte value. Cast the value to the wider type before calling operator '*' to avoid overflow (io.2).
47. PREfast Note C26451 - File: onnxruntime/contrib_ops/cpu/bert/attention_cpu_base.h. Line: 219. Column 18. 
Tool: PREfast: Rule: C26451 
Arithmetic overflow: Using operator '*' on a 4 byte value and then casting the result to a 8 byte value. Cast the value to the wider type before calling operator '*' to avoid overflow (io.2).
48. PREfast Note C26451 - File: onnxruntime/contrib_ops/cpu/bert/attention_cpu_base.h. Line: 225. Column 36. 
Tool: PREfast: Rule: C26451 
Arithmetic overflow: Using operator '*' on a 4 byte value and then casting the result to a 8 byte value. Cast the value to the wider type before calling operator '*' to avoid overflow (io.2).
49. PREfast Note C26451 - File: onnxruntime/contrib_ops/cpu/bert/attention_helper.h. Line: 111. Column 48. 
Tool: PREfast: Rule: C26451 
Arithmetic overflow: Using operator '*' on a 4 byte value and then casting the result to a 8 byte value. Cast the value to the wider type before calling operator '*' to avoid overflow (io.2).
50. PREfast Note C26451 - File: onnxruntime/contrib_ops/cpu/bert/attention_helper.h. Line: 136. Column 23. 
Tool: PREfast: Rule: C26451 
Arithmetic overflow: Using operator '*' on a 4 byte value and then casting the result to a 8 byte value. Cast the value to the wider type before calling operator '*' to avoid overflow (io.2).
51. PREfast Note C26451 - File: onnxruntime/contrib_ops/cpu/bert/attention_helper.h. Line: 148. Column 15. 
Tool: PREfast: Rule: C26451 
Arithmetic overflow: Using operator '*' on a 4 byte value and then casting the result to a 8 byte value. Cast the value to the wider type before calling operator '*' to avoid overflow (io.2).
52. PREfast Note C26451 - File: onnxruntime/contrib_ops/cpu/transformers/generation_device_helper.cc. Line: 790. Column 24. 
Tool: PREfast: Rule: C26451 
Arithmetic overflow: Using operator '+' on a 4 byte value and then casting the result to a 8 byte value. Cast the value to the wider type before calling operator '+' to avoid overflow (io.2).
53. PREfast Note C26451 - File: onnxruntime/contrib_ops/cpu/transformers/generation_device_helper.cc. Line: 789. Column 19. 
Tool: PREfast: Rule: C26451 
Arithmetic overflow: Using operator '+' on a 4 byte value and then casting the result to a 8 byte value. Cast the value to the wider type before calling operator '+' to avoid overflow (io.2).
54. PREfast Note C26451 - File: onnxruntime/contrib_ops/cpu/transformers/generation_device_helper.cc. Line: 514. Column 44. 
Tool: PREfast: Rule: C26451 
Arithmetic overflow: Using operator '+' on a 4 byte value and then casting the result to a 8 byte value. Cast the value to the wider type before calling operator '+' to avoid overflow (io.2).
55. PREfast Note C26451 - File: onnxruntime/contrib_ops/cpu/transformers/generation_device_helper.cc. Line: 541. Column 17. 
Tool: PREfast: Rule: C26451 
Arithmetic overflow: Using operator '+' on a 4 byte value and then casting the result to a 8 byte value. Cast the value to the wider type before calling operator '+' to avoid overflow (io.2).
56. PREfast Note C26451 - File: onnxruntime/contrib_ops/cpu/transformers/generation_device_helper.cc. Line: 702. Column 44. 
Tool: PREfast: Rule: C26451 
Arithmetic overflow: Using operator '+' on a 4 byte value and then casting the result to a 8 byte value. Cast the value to the wider type before calling operator '+' to avoid overflow (io.2).
57. PREfast Note C26451 - File: onnxruntime/contrib_ops/cpu/transformers/generation_device_helper.cc. Line: 722. Column 17. 
Tool: PREfast: Rule: C26451 
Arithmetic overflow: Using operator '+' on a 4 byte value and then casting the result to a 8 byte value. Cast the value to the wider type before calling operator '+' to avoid overflow (io.2).
58. PREfast Note C26409 - File: onnxruntime/core/platform/windows/env.cc. Line: 514. Column 85. 
Tool: PREfast: Rule: C26409 
Avoid calling new and delete explicitly, use std::make_unique<T> instead (r.11).
59. PREfast Note C26497 - File: onnxruntime/core/providers/cpu/signal/dft.cc. Line: 43. Column 13. 
Tool: PREfast: Rule: C26497 
You can attempt to make 'onnxruntime::is_power_of_2' constexpr unless it contains any undefined behavior (f.4).
60. PREfast Note C26814 - File: onnxruntime/core/providers/cpu/signal/dft.cc. Line: 430. Column 37. 
Tool: PREfast: Rule: C26814 
The const variable 'output_components' can be computed at compile-time. Consider using constexpr (con.5).
61. PREfast Note C26451 - File: onnxruntime/core/providers/cpu/signal/dft.cc. Line: 415. Column 33. 
Tool: PREfast: Rule: C26451 
Arithmetic overflow: Using operator '+' on a 4 byte value and then casting the result to a 8 byte value. Cast the value to the wider type before calling operator '+' to avoid overflow (io.2).
62. PREfast Note C26814 - File: onnxruntime/core/providers/cpu/signal/dft.cc. Line: 82. Column 23. 
Tool: PREfast: Rule: C26814 
The const variable '`onnxruntime::compute_angular_velocity<double>'::`2'::pi' can be computed at compile-time. Consider using constexpr (con.5).
63. PREfast Note C26814 - File: onnxruntime/core/providers/cpu/signal/dft.cc. Line: 83. Column 26. 
Tool: PREfast: Rule: C26814 
The const variable '`onnxruntime::compute_angular_velocity<double>'::`2'::tau' can be computed at compile-time. Consider using constexpr (con.5).
64. PREfast Note C26451 - File: onnxruntime/core/providers/cpu/tensor/concat.cc. Line: 113. Column 23. 
Tool: PREfast: Rule: C26451 
Arithmetic overflow: Using operator '+' on a 4 byte value and then casting the result to a 8 byte value. Cast the value to the wider type before calling operator '+' to avoid overflow (io.2).
65. PREfast Note C26814 - File: onnxruntime/test/providers/cpu/signal/signal_ops_test.cc. Line: 129. Column 31. 
Tool: PREfast: Rule: C26814 
The const variable 'num_batches' can be computed at compile-time. Consider using constexpr (con.5).

### CODE REVIEW

File: onnxruntime/contrib_ops/cpu/bert/attention_base.cc [hunk 1]
```diff
@@ -375,7 +375,7 @@ Tensor* AttentionBase::GetPresent(OpKernelContext* context,
   //   present     : (2, batch_size, num_heads, past_sequence_length + kv_sequence_length, head_size)
 
   past_sequence_length = (nullptr != past) ? static_cast<int>(past->Shape().GetDims()[3]) : 0;
-  std::vector<int64_t> present_dims{2, batch_size, num_heads_, kv_sequence_length + past_sequence_length, head_size};
+  std::array<int64_t, 5> present_dims{2, batch_size, num_heads_, static_cast<int64_t>(kv_sequence_length) + past_sequence_length, head_size};
```
Comment (snnn) [inline onnxruntime/contrib_ops/cpu/bert/attention_base.cc:378]: Here both kv_sequence_length and past_sequence_length are `int`.  But the result will be converted to int64_t. So we can convert one of them to int64_t first to get rid of the chance of arithmetic overflow.

File: onnxruntime/contrib_ops/cpu/bert/attention_cpu_base.h [hunk 2]
```diff
@@ -207,22 +207,22 @@ class AttentionCPUBase : public AttentionBase {
                                T* present,                // present state
                                ThreadPool* tp) const {
     const int total_sequence_length = past_sequence_length + kv_sequence_length;               // T = P + L
-    const size_t past_chunk_length = static_cast<size_t>(past_sequence_length * v_head_size);  // P x H_v
-    const size_t input_chunk_length = static_cast<size_t>(kv_sequence_length * v_head_size);   // L x H_v
-    const size_t present_chunk_length = past_chunk_length + input_chunk_length;                // T x H_v
+    const ptrdiff_t past_chunk_length = SafeInt<ptrdiff_t>(past_sequence_length) * v_head_size;  // P x H_v
```
Comment (snnn) [inline onnxruntime/contrib_ops/cpu/bert/attention_cpu_base.h:210]: If we know past_sequence_length and v_head_size are non-negative, then we can set the type of past_chunk_length to `size_t`. But if we don't know, ptrdiff_t is safer.

File: onnxruntime/test/framework/session_state_test.cc [hunk 3]
```diff
@@ -530,249 +530,253 @@ TEST_P(SessionStatePrepackingTest, PrePackingTest) {
...
-TEST(SessionStateTest, SharedInitalizersWithPrePackingTest) {
-  OrtThreadPoolParams to;
-  auto tp = concurrency::CreateThreadPool(&onnxruntime::Env::Default(), to, concurrency::ThreadPoolType::INTRA_OP);
-  ONNX_OPERATOR_SCHEMA(PrePackingTest)
-      .SetDoc("Faking Node for PrePacking")
-      .Input(0, "Input_0", "input 0", "tensor(float)")
-      .Input(1, "Input_1", "input 1", "tensor(float)")
-      .Output(0, "output_0", "docstr for output_0.", "tensor(float)");
-
+class SessionStateTestSharedInitalizersWithPrePacking : public ::testing::Test {
```
Comment (snnn) [inline onnxruntime/test/framework/session_state_test.cc:533]: The original test function is too big. It uses too much stack memory.

File: onnxruntime/test/optimizer/graph_transform_test_builder.cc [hunk 4]
```diff
@@ -109,15 +109,15 @@ void TestGraphTransformer(const std::function<void(ModelTestBuilder& helper)>& b
               domain_to_version, {}, logger);
   Graph& graph = model.MainGraph();
   ModelTestBuilder helper(graph);
-  ASSERT_TRUE(build_test_case);
```
Comment (snnn) [inline onnxruntime/test/optimizer/graph_transform_test_builder.cc:112]: Such ASSERT macros abort the **current** function, it has no effect on the callers of this function.  And the old version this function returns 'void', so its callers has no information on whether it was successful or not.

File: onnxruntime/contrib_ops/cpu/bert/attention_base.cc (hunk 1 above)
Comment (jchen351) [reply onnxruntime/contrib_ops/cpu/bert/attention_base.cc:378]: Good practice to use static array.

**Review by jchen351 — APPROVED** [review]

**Review by yufenglee — APPROVED** [review]
:shipit:

**Pull request merged**